nntrainer/tensor/manager.cpp

```diff
@@ -245,40 +250,87 @@ void Manager::initialize() {
  * still derivative memory needs to be allocated
  */
 std::vector<std::shared_ptr<Var_Grad>> &
-Manager::TrackLayerInOuts(const std::string &layer_type,
+Manager::trackLayerInOuts(const std::string &layer_type,
                           const std::string &layer_name,
-                          const std::vector<TensorDim> &input_dim) {
+                          const std::vector<TensorDim> &inout_dim) {
   int cnt = 0;
-  auto base_name = layer_name + ":InOut";
   bool is_act_layer = layer_type == ActivationLayer::type;
+  bool is_flat_layer = layer_type == FlattenLayer::type;
 
   size_t inout_derivative_size = 0;
 
   std::vector<std::shared_ptr<Var_Grad>> in_out;
-  in_out.reserve(input_dim.size());
+  in_out.reserve(inout_dim.size());
 
-  for (auto const &dim : input_dim) {
-    in_out.emplace_back(
-      std::make_shared<Var_Grad>(dim, true, base_name + std::to_string(cnt++)));
+  for (auto const &dim : inout_dim) {
+    in_out.emplace_back(std::make_shared<Var_Grad>(
+      dim, true, layer_name + std::to_string(cnt++)));
     if (is_act_layer)
       inout_derivative_size += dim.getDataLen();
   }
 
   in_outs.push_back(in_out);
   is_act_type.push_back(is_act_layer);
+  is_flat_type.push_back(is_flat_layer);
 
   max_derivative_size = std::max(max_derivative_size, inout_derivative_size);
   return in_outs.back();
 }
 
+std::vector<std::shared_ptr<Var_Grad>> &
+Manager::trackLayerOutputs(const std::string &layer_type,
+                           const std::string &layer_name,
+                           const std::vector<TensorDim> &output_dim,
+                           const std::vector<TensorDim> &input_dim) {
+  if (enable_inference_inout_memory_opt && input_dim.empty()) {
+    throw std::invalid_argument(
+      "Input dimensions are required with inference memory opt.");
+  } else {
```

Comment:
Could you double check this if else statement?

what if `enable_inference_inout_memory_opt` is true && `!input_dim.empty()` vs `enable_inference_inout_memory_opt` is false

Reply:
`if enable_inference_inout_memory_opt is true && !input_dim.empty()` -> then use input_dim to estimate the required memory, and then track layer inouts.
`if enable_inference_inout_memory_opt is false` -> added an extra check to estimate required memory when enable_inference_inout_memory_opt is true.

Reply:
awesome